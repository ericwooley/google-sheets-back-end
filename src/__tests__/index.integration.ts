// @jest-environment browser
import * as fs from "fs"
import { join } from "path"
import initSheets, { IHaveSheets, removeSheetJunk } from ".."
import wineSchema from "./mocks/wineSchema"

const TOKEN: string = process.env.AUTH_TOKEN || ""
function updateMock(file: string, data: string) {
  if (process.env.UPDATE_MOCKS) {
    fs.writeFileSync(join(__dirname, "./mocks/", file), data)
  }
}
const removeAndTestChaningSpreadSheetProps = <T extends IHaveSheets>(
  sheetData: T
): T => {
  return {
    ...Object.assign(sheetData),
    sheets: sheetData.sheets.map((sheet, index) => {
      expect(sheet.properties.sheetId).toBeDefined()
      return {
        ...sheet,
        properties: {
          ...sheet.properties,
          sheetId: `<sheetId for ${index}>` as any
        }
      }
    })
  }
}
const makeSheetDataSerializable = <T extends IHaveSheets>(sheetData: T) =>
  removeSheetJunk(removeAndTestChaningSpreadSheetProps(sheetData))
describe("Google Sheets As a db", () => {
  beforeAll(() => {
    if (!TOKEN) {
      throw new Error(
        "Integration tests require an oauth token from the google oauth2 playground"
      )
    }
    if (TOKEN.length < 100) {
      console.warn(`Your token appears to be too short ${TOKEN.length}`)
    }
  })
  describe("initializing as a database", () => {
    it("should initialize", () => {
      const sheets = initSheets({
        accessToken: TOKEN
      })
      expect(sheets).toMatchSnapshot("returned-api")
    })
  })
  describe("auth error", () => {
    it("should throw an error if a request failes", async () => {
      const sheets = initSheets({
        accessToken: "<TEST>"
      })
      try {
        await sheets.createDB({
          title: "wahtever",
          schema: wineSchema
        })
      } catch (e) {
        expect(e.message).toMatchSnapshot()
      }
    })
  })
  describe("Getting a sheet", () => {
    let sheets = initSheets({
      accessToken: TOKEN
    })
    beforeEach(() => {
      sheets = initSheets({
        accessToken: TOKEN
      })
    })
    it("should get a sheet", async () => {
      const result = await sheets.getSheet(
        "1sQpusd4BmF3kbTaTcdkOSYihSD0ltVnOUn_bKY14i7Y"
      )
      updateMock("template_sheet.json", JSON.stringify(result, null, 2))

      expect(makeSheetDataSerializable(result)).toMatchSnapshot()
    })
  })
  describe("Using as a database", () => {
    const sheets = initSheets({
      accessToken: TOKEN
    })
    describe("create a new db", () => {
      it("should create a new Database", async () => {
        const db = await sheets.createDB({
          title: "test-sheet" + Date.now(),
          schema: wineSchema
        })
        expect(
          makeSheetDataSerializable({ sheets: [db.getEntity("wine")._raw()] })
        ).toMatchSnapshot()
      })
    })
  })
})
