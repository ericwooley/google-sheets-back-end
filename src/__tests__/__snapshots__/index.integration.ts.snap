// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google Sheets As a db Getting a sheet should get a sheet 1`] = `
Object {
  "properties": Object {
    "autoRecalc": "ON_CHANGE",
    "defaultFormat": Object {
      "backgroundColor": Object {
        "blue": 1,
        "green": 1,
        "red": 1,
      },
      "padding": Object {
        "bottom": 2,
        "left": 3,
        "right": 3,
        "top": 2,
      },
      "textFormat": Object {
        "bold": false,
        "fontFamily": "arial,sans,sans-serif",
        "fontSize": 10,
        "foregroundColor": Object {},
        "italic": false,
        "strikethrough": false,
        "underline": false,
      },
      "verticalAlignment": "BOTTOM",
      "wrapStrategy": "OVERFLOW_CELL",
    },
    "locale": "en_US",
    "timeZone": "America/Los_Angeles",
    "title": "Malias Wine Sheet Template",
  },
  "sheets": Array [
    Object {
      "data": Array [
        Object {
          "columnMetadata": Array [],
          "rowData": Array [
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                  "formattedValue": "Do not edit column names!",
                  "userEnteredValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
              ],
            },
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "id",
                  },
                  "formattedValue": "id",
                  "userEnteredValue": Object {
                    "stringValue": "id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "name",
                  },
                  "formattedValue": "name",
                  "userEnteredValue": Object {
                    "stringValue": "name",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "vintage",
                  },
                  "formattedValue": "vintage",
                  "userEnteredValue": Object {
                    "stringValue": "vintage",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "producer",
                  },
                  "formattedValue": "producer",
                  "userEnteredValue": Object {
                    "stringValue": "producer",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "vineyardDesignate",
                  },
                  "formattedValue": "vineyardDesignate",
                  "userEnteredValue": Object {
                    "stringValue": "vineyardDesignate",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "subAppellation",
                  },
                  "formattedValue": "subAppellation",
                  "userEnteredValue": Object {
                    "stringValue": "subAppellation",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkBefore",
                  },
                  "formattedValue": "drinkBefore",
                  "userEnteredValue": Object {
                    "stringValue": "drinkBefore",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkAfter",
                  },
                  "formattedValue": "drinkAfter",
                  "userEnteredValue": Object {
                    "stringValue": "drinkAfter",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkDate",
                  },
                  "formattedValue": "drinkDate",
                  "userEnteredValue": Object {
                    "stringValue": "drinkDate",
                  },
                },
              ],
            },
          ],
          "rowMetadata": Array [],
        },
      ],
      "properties": Object {
        "gridProperties": Object {
          "columnCount": 26,
          "rowCount": 1000,
        },
        "index": 0,
        "sheetId": "<sheetId for 0>",
        "sheetType": "GRID",
        "title": "Wines",
      },
    },
    Object {
      "data": Array [
        Object {
          "columnMetadata": Array [],
          "rowData": Array [
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                  "formattedValue": "Do not edit column names!",
                  "userEnteredValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
              ],
            },
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "id",
                  },
                  "formattedValue": "id",
                  "userEnteredValue": Object {
                    "stringValue": "id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "cellarId",
                  },
                  "formattedValue": "cellarId",
                  "userEnteredValue": Object {
                    "stringValue": "cellarId",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "wineId",
                  },
                  "formattedValue": "wineId",
                  "userEnteredValue": Object {
                    "stringValue": "wineId",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "row",
                  },
                  "formattedValue": "row",
                  "userEnteredValue": Object {
                    "stringValue": "row",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "column",
                  },
                  "formattedValue": "column",
                  "userEnteredValue": Object {
                    "stringValue": "column",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "depth",
                  },
                  "formattedValue": "depth",
                  "userEnteredValue": Object {
                    "stringValue": "depth",
                  },
                },
              ],
            },
          ],
          "rowMetadata": Array [],
        },
      ],
      "properties": Object {
        "gridProperties": Object {
          "columnCount": 29,
          "rowCount": 1000,
        },
        "index": 1,
        "sheetId": "<sheetId for 1>",
        "sheetType": "GRID",
        "title": "location",
      },
    },
    Object {
      "data": Array [
        Object {
          "columnMetadata": Array [],
          "rowData": Array [
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                  "formattedValue": "Do not edit column names!",
                  "userEnteredValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
              ],
            },
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "id",
                  },
                  "formattedValue": "id",
                  "userEnteredValue": Object {
                    "stringValue": "id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "name",
                  },
                  "formattedValue": "name",
                  "userEnteredValue": Object {
                    "stringValue": "name",
                  },
                },
              ],
            },
          ],
          "rowMetadata": Array [],
        },
      ],
      "properties": Object {
        "gridProperties": Object {
          "columnCount": 26,
          "rowCount": 1000,
        },
        "index": 2,
        "sheetId": "<sheetId for 2>",
        "sheetType": "GRID",
        "title": "cellars",
      },
    },
    Object {
      "data": Array [
        Object {
          "columnMetadata": Array [],
          "rowData": Array [
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                  "formattedValue": "Do not edit column names!",
                  "userEnteredValue": Object {
                    "stringValue": "Do not edit column names!",
                  },
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
                Object {
                  "effectiveValue": undefined,
                  "formattedValue": undefined,
                  "userEnteredValue": undefined,
                },
              ],
            },
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "id",
                  },
                  "formattedValue": "id",
                  "userEnteredValue": Object {
                    "stringValue": "id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "date",
                  },
                  "formattedValue": "date",
                  "userEnteredValue": Object {
                    "stringValue": "date",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "wine_id",
                  },
                  "formattedValue": "wine_id",
                  "userEnteredValue": Object {
                    "stringValue": "wine_id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "eventName",
                  },
                  "formattedValue": "eventName",
                  "userEnteredValue": Object {
                    "stringValue": "eventName",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "eventNotes",
                  },
                  "formattedValue": "eventNotes",
                  "userEnteredValue": Object {
                    "stringValue": "eventNotes",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "wineRating",
                  },
                  "formattedValue": "wineRating",
                  "userEnteredValue": Object {
                    "stringValue": "wineRating",
                  },
                },
              ],
            },
          ],
          "rowMetadata": Array [],
        },
      ],
      "properties": Object {
        "gridProperties": Object {
          "columnCount": 27,
          "rowCount": 1001,
        },
        "index": 3,
        "sheetId": "<sheetId for 3>",
        "sheetType": "GRID",
        "title": "Log",
      },
    },
  ],
  "spreadsheetId": "1sQpusd4BmF3kbTaTcdkOSYihSD0ltVnOUn_bKY14i7Y",
  "spreadsheetUrl": "https://docs.google.com/spreadsheets/d/1sQpusd4BmF3kbTaTcdkOSYihSD0ltVnOUn_bKY14i7Y/edit",
}
`;

exports[`Google Sheets As a db Using as a database create a new db should create a new Database 1`] = `
Object {
  "sheets": Array [
    Object {
      "data": Array [
        Object {
          "columnMetadata": Array [],
          "rowData": Array [
            Object {
              "values": Array [
                Object {
                  "effectiveValue": Object {
                    "stringValue": "id",
                  },
                  "formattedValue": "id",
                  "userEnteredValue": Object {
                    "stringValue": "id",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "name",
                  },
                  "formattedValue": "name",
                  "userEnteredValue": Object {
                    "stringValue": "name",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "vintage",
                  },
                  "formattedValue": "vintage",
                  "userEnteredValue": Object {
                    "stringValue": "vintage",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "producer",
                  },
                  "formattedValue": "producer",
                  "userEnteredValue": Object {
                    "stringValue": "producer",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "vineyardDesignate",
                  },
                  "formattedValue": "vineyardDesignate",
                  "userEnteredValue": Object {
                    "stringValue": "vineyardDesignate",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "subAppellation",
                  },
                  "formattedValue": "subAppellation",
                  "userEnteredValue": Object {
                    "stringValue": "subAppellation",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkBefore",
                  },
                  "formattedValue": "drinkBefore",
                  "userEnteredValue": Object {
                    "stringValue": "drinkBefore",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkAfter",
                  },
                  "formattedValue": "drinkAfter",
                  "userEnteredValue": Object {
                    "stringValue": "drinkAfter",
                  },
                },
                Object {
                  "effectiveValue": Object {
                    "stringValue": "drinkDate",
                  },
                  "formattedValue": "drinkDate",
                  "userEnteredValue": Object {
                    "stringValue": "drinkDate",
                  },
                },
              ],
            },
          ],
          "rowMetadata": Array [],
        },
      ],
      "properties": Object {
        "gridProperties": Object {
          "columnCount": 26,
          "rowCount": 1000,
        },
        "index": 0,
        "sheetId": "<sheetId for 0>",
        "sheetType": "GRID",
        "title": "wine",
      },
    },
  ],
}
`;

exports[`Google Sheets As a db auth error should throw an error if a request failes 1`] = `"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."`;

exports[`returned-api 1`] = `
Object {
  "copySheet": [Function],
  "createDB": [Function],
  "createSheet": [Function],
  "getSheet": [Function],
  "loadDB": [Function],
}
`;
